xc/device.ml

```diff
@@ -2065,8 +2065,29 @@ module Backend = struct
 
       let suspend (task: Xenops_task.task_handle) ~xs ~qemu_domid domid =
         Dm_Common.suspend task ~xs ~qemu_domid domid;
-        let file = sprintf qemu_save_path domid in
-        qmp_write domid (Qmp.Command(None, Qmp.Xen_save_devices_state file))
+        let open Qmp in
+        let c = Qmp_protocol.connect (qmp_libxl_path domid) in
+        finally
+          (fun () ->
+             let fd_of_c = Qmp_protocol.to_fd c in
+             let fd_of_save_file = Unix.openfile (sprintf qemu_save_path domid) [ Unix.O_WRONLY; Unix.O_CREAT ] 0o660 in
+             finally
+               (fun () -> ignore(Fd_send_recv.send_fd fd_of_c " " 0 1 [] fd_of_save_file))
+               (fun () -> Unix.close fd_of_save_file);
+
+             let qmp_cmd = Command (None, Add_fd None) in
+             Qmp_protocol.negotiate c;
+             Qmp_protocol.write c qmp_cmd;
+             let fd_info = match Qmp_protocol.read c with
+               | Success (None, Fd_info x) -> x
+               | _ -> raise (Internal_error (Printf.sprintf "Get unexpected result after sending Qmp message: %s" (string_of_message qmp_cmd)))
+             in
+             finally
+               (fun () ->
+                  Qmp_protocol.write c (Qmp.Command (None, Xen_save_devices_state ("/dev/fdset/" ^ (string_of_int fd_info.fdset_id)))))
```

Comment:
Since module `Qmp` is already open, it should be enough to use `(Command (None, ...))`. The same applies below. I probably would have used `sprintf "/dev/fdset/%d"  fd_info.fdset_id`. To avoid long lines, you could let-bind the command like it is done above.